### DISCUSSION TITLE
What's the deal with UNW_INFO_FORMAT_DYNAMIC?

### DISCUSSION THREAD

**Discussion by KJTsanaktsidis** [libunwind/libunwind#631] [Ideas]
I'm experimenting with adding unwinding support to a run-time code generator which targets x86_64 & aarch64 on Linux [1]. The stacks it generates are pretty simple at the moment (it generates frame pointers 100% of the time, for example). The code genreator is very dynamic though - it lazily generates code for blocks as they're executed and has a code GC strategy for cleaning up code when the executable memory mapping gets full.

These patterns make registering full-blown DWARF CFI unwinding info a bit annoying - the prologues & epilogues are very simple, so a full blown DWARF generator implementation isn't really needed. The requirement that a synthesied `.eh_frame` section be contiguous and sorted is also very inconvenient in the face of the code GC (the `.eh_frame` section would need to be rearranged & resorted when code is inserted in the middle of the mapping).

It seems on paper that registering dynamic info with libunwind using `UNW_INFO_FORMAT_DYNAMIC` would be a perfect fit; we don't need any more complex stack manipulation directives other than the ones documented in `libunwind-dynamic(3)`. Plus, we can nicely shrae a pointer to a single `unw_dyn_info_region_t` for our common prologues & epilogues, and the use of pointers for this info means we can easily free it when a routine gets GC'd.

The catch: `UNW_INFO_FORMAT_DYNAMIC` is not implemented for anything but Itanium :o

* This is only sort of a problem on x86_64, because there libunwind does fall back to frame pointer chasing, which "mostly" works (except when unwinding from the prologue/epilogue, of course)
* This is more of a problem on aarch64, because AFAICT libunwind does _not_ fall back to frame pointer chasing on this arch.
 
So, my questions:

* What's the reason why this format hasn't been implemented for any other architecture?
* Would a contribution be welcome here? Or is the dynamic unwind format just a bad idea for some reason?
* if it is a bad idea and you don't want a contribution, can the docs perhaps be updated to prominently note it's only implemented for Itanium?

Thanks!

[1] - The code generator in question is Ruby's YJIT jit compiler & there's a bit more context here: https://bugs.ruby-lang.org/issues/19541 - note that I'm not actually affiliated with the YJIT team in any way, it's a contribution I'm trying to make.

**Comment by bregma** [comment]
A contribution would be welcome.

**Comment by KJTsanaktsidis** [comment]
OK, I'll see what I can come up with!

**Comment by KJTsanaktsidis** [comment]
https://github.com/libunwind/libunwind/pull/495 - I took a stab at a spike for this, WDYT?
